### PR TITLE
ARROW-12133: [C++][Gandiva] Add option to disable targeting host cpu during llvm ir compilation

### PR DESCRIPTION
**Pull request by projjal**
Because of the issue mentioned in #9852 , it seems like a good idea to allow a configuration parameter to disable targeting host cpu during jit compilation if there are some unforseen crashes (in case llvm::sys::getHostCPUFeatures has a bug and returns incorrect info)

### CODE REVIEW

**Comment by projjal** [comment]
@praveenbingo @pravindra can you review?

File: cpp/src/gandiva/engine.cc [hunk 1]
```diff
@@ -133,13 +132,17 @@ Status Engine::Make(const std::shared_ptr<Configuration>& conf,
...
-          .create()};
+
+  llvm::EngineBuilder engine_builder(std::move(module));
+
+  engine_builder.setEngineKind(llvm::EngineKind::JIT)
+      .setOptLevel(opt_level)
+      .setErrorStr(&builder_error);
+
+  if (conf->target_host_cpu()) {
+    engine_builder.setMCPU(llvm::sys::getHostCPUName());
```
Comment (vvellanki) [inline cpp/src/gandiva/engine.cc:143]: What is the impact of not setting this? Does LLVM generate sub-optimal code? or does it internally figure out the CPU and generate appropriate code for that?

Comment (vvellanki) [reply cpp/src/gandiva/engine.cc:143]: This change conflicts with the other change as the call to getHostCPUName() is done once in the other change

File: java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/JniLoader.java [hunk 2]
```diff
@@ -35,7 +37,8 @@
 
   private static volatile JniLoader INSTANCE;
   private static volatile long defaultConfiguration = 0L;
-  private static volatile long unoptimizedConfiguration = 0L;
+  private static final ConcurrentMap<ConfigurationBuilder.ConfigOptions, Long> configurationMap
```
Comment (vvellanki) [inline java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/JniLoader.java:40]: Do we have tests to ensure that this map works as expected?

Using the same Config should return the same long. Release should work as expected, etc.

Comment (vvellanki) [reply java/gandiva/src/main/java/org/apache/arrow/gandiva/evaluator/JniLoader.java:40]: I see a releaseConfig in JNI. Dont we need something similar here to reduce the size of the Map?

File: cpp/src/gandiva/engine.cc (hunk 1 above)
Comment (projjal) [reply cpp/src/gandiva/engine.cc:143]: It won't use the cpu specific features (eg avx instructions)

Comment (projjal) [reply cpp/src/gandiva/engine.cc:143]: The conflict is mentioned in the PR description. One needs to be rebased when the other is merged.

**Review by vvellanki — APPROVED** [review]

**Review by praveenbingo — APPROVED** [review]

**Pull request closed without merging**